### PR TITLE
StudentsLogic: Simplify createStudents(String, String) #7839

### LINKED ISSUE #7839: StudentsLogic: Simplify createStudents(String, String)
The code is too hard to understand. 

```
if (StringHelper.isWhiteSpace(line)) {
    continue;
}
```
This code should be called immediately after the for-loop instead of the try-block. 

We can also look to remove the variable `studentEmailList` and simplify this chunk of code to use `studentList` instead:
```
if (isStudentEmailDuplicated(student.email, studentEmailList)) {
    String info =
            StringHelper.toString(
                    getInvalidityInfoInDuplicatedEmail(student.email, studentEmailList, linesArray),
                    "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ");
    invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
}
```

We can also try to apply Single Level of Abstraction Principle (SLAP) on this method as it is too long.

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #7839

### CODE REVIEW

**Comment by sujeet14108** [comment]
Ready for review :)

**Comment by whipermr5** [comment]
@Zhiyuan-Amos How's this?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 1]
```diff
@@ -629,15 +627,26 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
         List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
+        int index;
+        for (index = 0; index < studentList.size(); index++) {
+            if (studentList.get(index).email.equalsIgnoreCase(email)) {
+                break;
+            }
+        }
+        info.add("Same email address as the student in line \"" + linesArray[index + 1] + "\"");
         return info;
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: do we need the `break` at line 635? Otherwise we can shift lines 638 & 639 there instead.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 2]
```diff
@@ -629,15 +627,26 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+            List<StudentAttributes> studentList, String[] linesArray) {
         List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
+        int index;
+        for (index = 0; index < studentList.size(); index++) {
+            if (studentList.get(index).email.equalsIgnoreCase(email)) {
+                break;
+            }
+        }
+        info.add("Same email address as the student in line \"" + linesArray[index + 1] + "\"");
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: is there a reason why are we using a `List` when we are only adding a maximum of 1 element?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 3]
```diff
@@ -629,15 +627,26 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
     private boolean isStudentEmailDuplicated(String email,
-            List<String> studentEmailList) {
-        return studentEmailList.contains(email);
+            List<StudentAttributes> studentList) {
+        for (StudentAttributes studentInfo : studentList) {
+            if (studentInfo.email.equalsIgnoreCase(email)) {
+                return true;
+            }
+        }
+        return false;
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: I noticed that the logic here is quite similar to `getInvalidityInfoInDuplicatedEmail`. 

@whipermr5 do you think it's good if we remove `isStudentEmailDuplicated` and replace it with something like `getDuplicateIndex(String email)`, which returns `-1` if `email` isn't a duplicate and returns the corresponding index if it's a duplicate. That way, we do not have duplicate logic.

**Comment by Zhiyuan-Amos** [comment]
@whipermr5 this method can be further SLAP-ed, though I'm not sure how clean you want it to be. ^^

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 1 above)
Comment (sujeet14108) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Yes can be done like this :)

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 3 above)
Comment (sujeet14108) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: @Zhiyuan-Amos I think we can remove `isStudentEmailDuplicated` . @whipermr5  what do you think ?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 4]
```diff
@@ -628,16 +625,27 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
         return studentList;
     }
 
-    private List<String> getInvalidityInfoInDuplicatedEmail(String email,
-            List<String> studentEmailList, String[] linesArray) {
-        List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
+    private String getInvalidityInfoInDuplicatedEmail(String email,
+            List<StudentAttributes> studentList, String[] linesArray) {
+        String info = null;
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: `info` isn't necessary. Remove line 635 and have the return statement on line 634 instead.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 5]
```diff
@@ -628,16 +625,27 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+        int index;
+        for (index = 0; index < studentList.size(); index++) {
+            if (studentList.get(index).email.equalsIgnoreCase(email)) {
+                info = "Same email address as the student in line \"" + linesArray[index + 1] + "\"";
+                return info;
+            }
+        }
         return info;
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: following the above comment, change this to `return ""`.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 6]
```diff
@@ -628,16 +625,27 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
 
-    private List<String> getInvalidityInfoInDuplicatedEmail(String email,
-            List<String> studentEmailList, String[] linesArray) {
-        List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
+    private String getInvalidityInfoInDuplicatedEmail(String email,
+            List<StudentAttributes> studentList, String[] linesArray) {
+        String info = null;
+        int index;
+        for (index = 0; index < studentList.size(); index++) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: bring the variable declaration down into the for loop

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 3 above)
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Just go ahead and we'll review it for you ^^

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 7]
```diff
@@ -585,18 +585,17 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+            if (StringHelper.isWhiteSpace(line)) {
+                continue;
+            }
             try {
-                if (StringHelper.isWhiteSpace(line)) {
-                    continue;
-                }
                 StudentAttributes student = saf.makeStudent(line, courseId);
 
                 if (!student.isValid()) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Let's SLAP this i.e. extract this into a helper method `addInvalidStudentInfo(List<String> invalidityInfo, StudentAttributes student, String sanitizedLine)`.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 8]
```diff
@@ -605,15 +604,13 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
                     invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
                 }
 
-                if (isStudentEmailDuplicated(student.email, studentEmailList)) {
+                if (isStudentEmailDuplicated(student.email, studentList)) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Let's SLAP this too, similar to what I've suggested above.

**Comment by Zhiyuan-Amos** [comment]
@whipermr5 I'm thinking of making the try-catch block smaller. What do you think?

```
StudentAttributes student = null;

try {
    if (StringHelper.isWhiteSpace(line)) {
        continue;
    }
    student = saf.makeStudent(line, courseId);
} catch (EnrollException e) {
    String info = String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, e.getMessage());
    invalidityInfo.add(info);
}

if (student == null) {
    continue;
}  

... // remaining code that were previously in the try-catch block
```

**Comment by sujeet14108** [comment]
@Zhiyuan-Amos Have a look :)

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 9]
```diff
@@ -628,16 +620,24 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+            }
+        }
+        return "";
     }
 
-    private boolean isStudentEmailDuplicated(String email,
-            List<String> studentEmailList) {
-        return studentEmailList.contains(email);
+    private int getDuplicateIndex(String email,
+            List<StudentAttributes> studentList) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: this line can be on the same line as the previous line?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 10]
```diff
@@ -585,35 +585,27 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
-                                    "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ");
-                    invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
+                if (getDuplicateIndex(student.email, studentList) != -1) {
+                    addDuplicateEmailInfo(invalidityInfo, student, sanitizedLine, studentList, linesArray);
                 }
 
-                studentEmailList.add(student.email);
                 studentList.add(student);
             } catch (EnrollException e) {
                 String info = String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, e.getMessage());
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: we can SLAP this too, since you've already SLAP-ed `addDuplicateEmailInfo` and `addInvalidStudentInfo` :)

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 11]
```diff
@@ -671,4 +671,18 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
         return null;
     }
 
+    private void addInvalidStudentInfo(List<String> invalidityInfo, StudentAttributes student, String sanitizedLine) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: can we bring these methods up to directly after `getDuplicateIndex` since these methods are related to each other.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 12]
```diff
@@ -585,35 +585,27 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
                 }
 
-                if (isStudentEmailDuplicated(student.email, studentEmailList)) {
-                    String info =
-                            StringHelper.toString(
-                                    getInvalidityInfoInDuplicatedEmail(student.email, studentEmailList, linesArray),
-                                    "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ");
-                    invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
+                if (getDuplicateIndex(student.email, studentList) != -1) {
+                    addDuplicateEmailInfo(invalidityInfo, student, sanitizedLine, studentList, linesArray);
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: You should store the return value of `getDuplicateIndex` and pass in the value to `addDuplicateEmailInfo`. That way, `getInvalidityInfoInDuplicatedEmail` does not have to the search for the index again.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 13]
```diff
@@ -628,16 +620,24 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+        return "";
     }
 
-    private boolean isStudentEmailDuplicated(String email,
-            List<String> studentEmailList) {
-        return studentEmailList.contains(email);
+    private int getDuplicateIndex(String email,
+            List<StudentAttributes> studentList) {
+        for (int index = 0; index < studentList.size(); index++) {
+            if (studentList.get(index).email.equalsIgnoreCase(email)) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: is there a reason for using `equalsIgnoreCase`? The previous code was something like `linesArray[studentEmailList.indexOf(email) + 1]`, and the method `indexOf` uses the `equals` method for comparison if I'm not wrong ^^

Comment (sujeet14108) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: I used `equalsIgnoreCase` because others previously defined methods like `isInEnrollList` have used this. 
So should i use `equal` instead ?

**Comment by sujeet14108** [comment]
@Zhiyuan-Amos Have a look :)

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 14]
```diff
@@ -628,20 +619,32 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+    private void addInvalidStudentInfo(List<String> invalidityInfo, StudentAttributes student, String sanitizedLine) {
+        String info = StringHelper.toString(SanitizationHelper.sanitizeForHtml(student.getInvalidityInfo()),
+                "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ");
+        invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
     }
 
-    private boolean isInEnrollList(StudentAttributes student,
-            List<StudentAttributes> studentInfoList) {
+    /**
+     * Check for the duplicate email, if duplicate is found returns its index in studentList else return -1.
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: We need better header comments for this :P Just use mine:
`Returns the index of the first occurrence of the duplicate {@code email} in {@code studentList}, or -1 if {@code email} is not a duplicate in {@code studentList}.`

Copied from ArrayList#indexOf(Object) :P

@whipermr5 you ok with this format?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 15]
```diff
@@ -585,39 +585,30 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+                    addInvalidStudentInfo(invalidityInfo, student, sanitizedLine);
                 }
-
-                if (isStudentEmailDuplicated(student.email, studentEmailList)) {
-                    String info =
-                            StringHelper.toString(
-                                    getInvalidityInfoInDuplicatedEmail(student.email, studentEmailList, linesArray),
-                                    "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ");
-                    invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
+                int isDuplicate = getDuplicateIndex(student.email, studentList);
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: let's add a new line above it for code readability.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 16]
```diff
@@ -628,20 +619,32 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
     }
 
-    private boolean isInEnrollList(StudentAttributes student,
-            List<StudentAttributes> studentInfoList) {
+    /**
+     * Check for the duplicate email, if duplicate is found returns its index in studentList else return -1.
+     */
+    private int getDuplicateIndex(String email, List<StudentAttributes> studentList) {
+        for (int index = 0; index < studentList.size(); index++) {
+            if (studentList.get(index).email.equalsIgnoreCase(email)) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: you should use `equals` instead as the previous code was `studentEmailList.contains(email);`, and `contains` use `equals(Object)` method.

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 14 above)
Comment (whipermr5) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Yup looks good! Thanks for helping with the reviews btw; really appreciate it!!

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 15 above)
Comment (sujeet14108) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Is `duplicateIndex` fine or simply `index` ?

Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Let's go with duplicateEmailIndex. 

The method should be renamed to getDuplicateEmailIndex for better clarity.

Comment (sujeet14108) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Done :)

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 17]
```diff
@@ -628,22 +620,32 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
         return studentList;
     }
 
-    private List<String> getInvalidityInfoInDuplicatedEmail(String email,
-            List<String> studentEmailList, String[] linesArray) {
-        List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
-        return info;
+    private String getInvalidityInfoInDuplicatedEmail(String[] linesArray, int duplicateEmailIndex) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: inline this method

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 18]
```diff
@@ -628,22 +620,32 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+        return "Same email address as the student in line \"" + linesArray[duplicateEmailIndex + 1] + "\"";
     }
 
-    private boolean isStudentEmailDuplicated(String email,
-            List<String> studentEmailList) {
-        return studentEmailList.contains(email);
+    private String addInvalidStudentInfo(StudentAttributes student, String sanitizedLine) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: can you leave a header comment to describe what is `sanitizedLine`? It isn't intuitive on first look. You can refer to the caller method `createStudents` as there's a `@param lines`, which is similar to `sanitizedLine`.

Also, you should actually rename `sanitizedLine` because you can't tell what this variable refers to from its naming. `sanitizedLine` refers to something like the details of the students or something like that isn't it.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 19]
```diff
@@ -671,4 +673,15 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
         return null;
     }
 
+    private String addDuplicateEmailInfo(String sanitizedLine, String[] linesArray, int isDuplicate) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: 1. Write header comments.
1. Try changing the method header to take in `String duplicateLine, String sanitizedLine` instead i.e. remove `String[] linesArray, int isDuplicate`. See if it looks better. We don't need to pass in the entire array if we are only using 1 element of the array.
1. Rename the method name: the caller is doing `invalidityInfo.add(addDuplicateEmailInfo(...));` which looks weird. Renaming it to `duplicateEmailInfo` sounds better.
1. Rename `sanitizedLine`.

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 18 above)
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Also, update the method name. The caller is doing `invalidityInfo.add(addInvalidStudentInfo(...));`, which doesn't read off well. Maybe something like `invalidStudentInfo(...)` will be better.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 20]
```diff
@@ -628,22 +620,32 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+        return String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info);
     }
 
-    private boolean isInEnrollList(StudentAttributes student,
-            List<StudentAttributes> studentInfoList) {
+    /**
+     * Returns the index of the first occurrence of the duplicate {@code email} in
+     * {@code studentList}, or -1 if {@code email} is not a duplicate in {@code studentList}.
+     */
+    private int getDuplicateIndex(String email, List<StudentAttributes> studentList) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Rename it to `getDuplicateEmailIndex` to make it clearer.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 21]
```diff
@@ -628,22 +620,32 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+                return index;
+            }
+        }
+        return -1;
+    }
+
+    private boolean isInEnrollList(StudentAttributes student, List<StudentAttributes> studentInfoList) {
         for (StudentAttributes studentInfo : studentInfoList) {
-            if (studentInfo.email.equalsIgnoreCase(student.email)) {
+            if (studentInfo.email.equals(student.email)) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: This line shouldn't be updated; you've updated the code of the wrong method. Refer to: https://github.com/TEAMMATES/teammates/pull/7930#discussion_r134698402

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 22]
```diff
@@ -671,4 +673,15 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
...
     }
 
+    private String addDuplicateEmailInfo(String sanitizedLine, String[] linesArray, int isDuplicate) {
+        String info =
+                getInvalidityInfoInDuplicatedEmail(linesArray, isDuplicate)
+                + "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ";
+        return String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info);
+    }
+
+    private String enrollExceptionInfo(String sanitizedLine, String errorMessage) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Write a header comment here as well: `sanitizedLine` is hard to understand.

**Comment by Zhiyuan-Amos** [comment]
Some issue haven't been addressed yet.

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 17 above)
Comment (sujeet14108) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: How inline method is written ? All in one line ? :sweat_smile:

Comment (Zhiyuan-Amos) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: Are you using IntelliJ? You can click on refactor -> inline and see what happens :P

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 23]
```diff
@@ -671,4 +676,21 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
         return null;
     }
 
+    /**
+     * lines is the enrollment lines entered by the instructor.
+     */
+    private String addDuplicateEmailInfo(String sanitizedLine, String[] linesArray, int isDuplicate) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: why is there the word `add` for this method, when the other helper methods do not have it?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 24]
```diff
@@ -628,20 +620,33 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+        return "Same email address as the student in line \"" + linesArray[duplicateEmailIndex + 1] + "\"";
     }
 
-    private boolean isStudentEmailDuplicated(String email,
-            List<String> studentEmailList) {
-        return studentEmailList.contains(email);
+    /**
+     * lines is the enrollment lines entered by the instructor.
+     */
+    private String invalidStudentInfo(StudentAttributes student, String line) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: 1. What I meant was, change the name of the variable `line` to something that is more descriptive, and update the header comments too. 

1. Also, to be more precise, this `line` isn't really the enrollment line entered by the instructor, because it has been sanitised already.

1. Follow the standard of header comments as per the one I gave you earlier, using annotations such as {@code} when denoting code, and also describing the return values.

1. You didn't specify what is this method returning.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 25]
```diff
@@ -677,17 +674,20 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
...
-     * lines is the enrollment lines entered by the instructor.
+     * Returns the formatted string for duplicate email error using {@code linesArray},
+     * the index of the first occurrence of the duplicate {@code isDuplicate}
+     * and sanitized enrollment line {@code line}.
      */
-    private String addDuplicateEmailInfo(String sanitizedLine, String[] linesArray, int isDuplicate) {
+    private String duplicateEmailInfo(String line, String[] linesArray, int isDuplicate) {
         String info =
-                getInvalidityInfoInDuplicatedEmail(linesArray, isDuplicate)
+                "Same email address as the student in line \"" + linesArray[isDuplicate + 1] + "\""
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: 1. Rename `isDuplicate`. I'm sure I mentioned it previously.

1. This String is a magic string: let's extract it out as a constant. As to where you should extract it out to (in this class? Or in `Const.StatusMessages`? Or elsewhere?), I'll leave it to you to figure it out.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 26]
```diff
@@ -620,12 +620,9 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
     }
 
-    private String getInvalidityInfoInDuplicatedEmail(String[] linesArray, int duplicateEmailIndex) {
-        return "Same email address as the student in line \"" + linesArray[duplicateEmailIndex + 1] + "\"";
-    }
-
     /**
-     * lines is the enrollment lines entered by the instructor.
+     * Returns the formatted string about enrollment problem for invalid using invalidity info of
+     * {@code student} and sanitized enrollment line {@code line}.
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: 1. Rename the variable {@code line} because it isn't descriptive. A line simply means `-----------` <- this thing is a line. Something like `userInput` will be more descriptive. Same for the other methods too.

1. Update header comments. 
<pre>
Returns <strike>the formatted string</strike> (the caller doesn't need to know 
that it's formatted or not) a {@code String} containing the invalid 
information of the {@code student} and the corresponding sanitised 
invalid {@code userInput}.
</pre>

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 25 above)
Comment (sujeet14108) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: Done

**Comment by sujeet14108** [comment]
@Zhiyuan-Amos Please review :)

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 27]
```diff
@@ -585,39 +585,31 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
 
-                if (isStudentEmailDuplicated(student.email, studentEmailList)) {
-                    String info =
-                            StringHelper.toString(
-                                    getInvalidityInfoInDuplicatedEmail(student.email, studentEmailList, linesArray),
-                                    "<br>" + Const.StatusMessages.ENROLL_LINES_PROBLEM_DETAIL_PREFIX + " ");
-                    invalidityInfo.add(String.format(Const.StatusMessages.ENROLL_LINES_PROBLEM, sanitizedLine, info));
+                int duplicateEmailIndex = getDuplicateEmailIndex(student.email, studentList);
+                if (duplicateEmailIndex != -1) {
+                    invalidityInfo.add(duplicateEmailInfo(sanitizedLine, linesArray, duplicateEmailIndex));
```
Comment (Zhiyuan-Amos) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Since we are only accessing an element in `linesArray`, we can simply pass in `linesArray[duplicateEmailIndex]`. 

Update the header comments as well.

Comment (sujeet14108) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: Have a look

**Comment by sujeet14108** [comment]
@whipermr5 Please review :)

**Comment by sujeet14108** [comment]
@whipermr5 Please review it

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 28]
```diff
@@ -628,20 +620,30 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
+            if (studentList.get(index).email.equals(email)) {
+                return index;
+            }
+        }
+        return -1;
     }
 
-    private boolean isInEnrollList(StudentAttributes student,
-            List<StudentAttributes> studentInfoList) {
+    private boolean isInEnrollList(StudentAttributes student, List<StudentAttributes> studentInfoList) {
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Unrelated change?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 29]
```diff
@@ -628,20 +620,30 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
-    private List<String> getInvalidityInfoInDuplicatedEmail(String email,
-            List<String> studentEmailList, String[] linesArray) {
-        List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
-        return info;
+    /**
+     * Returns a {@code String} containing the invalid information of the {@code student}
+     * and the corresponding sanitized invalid {@code userInput}.
+     */
+    private String invalidStudentInfo(StudentAttributes student, String userInput) {
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Can rearrange the parameters to `invalidStudentInfo(String userInput, StudentAttributes student)` for consistency with `duplicateEmailInfo` and `enrollExceptionInfo`?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 30]
```diff
@@ -671,4 +673,23 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
         return null;
     }
 
+    /**
+     * Returns a {@code String} containing the duplicate email information in {@code duplicateEmailInfo} and
+     * the corresponding sanitized invalid {@code userInput}.
+     */
+    private String duplicateEmailInfo(String userInput, String duplicateEmailInfo) {
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Can these two methods (this and `enrollExceptionInfo`) be moved to just after `getDuplicateEmailIndex`? This is more logical as it reflects the order in which the methods are used.

Comment (sujeet14108) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: Done

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 29 above)
Comment (sujeet14108) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: Done

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 31]
```diff
@@ -670,5 +692,4 @@ public TeamDetailsBundle getTeamDetailsForStudent(StudentAttributes student) {
         }
         return null;
     }
-
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Unrelated change.

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 32]
```diff
@@ -628,16 +620,46 @@ private StudentEnrollDetails enrollStudent(StudentAttributes validStudentAttribu
...
-    private List<String> getInvalidityInfoInDuplicatedEmail(String email,
-            List<String> studentEmailList, String[] linesArray) {
-        List<String> info = new ArrayList<>();
-        info.add("Same email address as the student in line \"" + linesArray[studentEmailList.indexOf(email) + 1] + "\"");
-        return info;
+    /**
+     * Returns a {@code String} containing the invalid information of the {@code student}
+     * and the corresponding sanitized invalid {@code userInput}.
+     */
+    private String invalidStudentInfo( String userInput, StudentAttributes student) {
```
Comment (whipermr5) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: This will probably cause build failure.

File: src/main/java/teammates/logic/core/StudentsLogic.java (hunk 31 above)
Comment (sujeet14108) [reply src/main/java/teammates/logic/core/StudentsLogic.java]: Reverted

**Comment by Zhiyuan-Amos** [comment]
On a side note @sujeet14108, do look through your PR before you request reviewers to review, because I noted that you made quite a number of minor mistakes every now and then, which could have been easily avoided if you looked through the changes again (for e.g., making unrelated changes, missing out some of the comments we've made earlier etc). 

This makes our review harder, so do take note of this in the future ^^ Though I must say that this PR definitely looks great now! :)
